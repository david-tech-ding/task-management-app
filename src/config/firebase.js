import { initializeApp } from "firebase/app";
import { getAuth } from "firebase/auth";

const firebaseConfig = {
  apiKey: "AIzaSyAJv-vJvf585MZ0poOjBWynBBiq9W4fCl4",
  authDomain: "task-management-app-d0b1d.firebaseapp.com",
  projectId: "task-management-app-d0b1d",
  storageBucket: "task-management-app-d0b1d.appspot.com",
  messagingSenderId: "357587538083",
  appId: "1:357587538083:web:2850e62cf8ae51bf87af16",
  measurementId: "G-YB9G43E89B",
};

const app = initializeApp(firebaseConfig);

export const auth = getAuth(app);

export default app;
