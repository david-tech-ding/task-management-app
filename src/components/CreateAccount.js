import React, { useState } from "react";
import axios from "axios";
import { useNavigate } from "react-router-dom";
import { createUserWithEmailAndPassword, updateProfile } from "firebase/auth";
import { auth } from "../config/firebase";
import "../styles/auth-form.css";

const CreateAccount = ({ onSetUser, onSetLoggedInUser, loggedInUser }) => {
  axios.defaults.baseURL = "http://localhost:3001";
  const [newUser, setNewUser] = useState({
    email: "",
    password: "",
    userName: "",
    firstName: "",
    lastName: "",
    jobRole: "",
  });
  const handleChange = (e) => {
    setNewUser({ ...newUser, [e.target.name]: e.target.value });
  };
  const navigate = useNavigate();
  const handleCreateAccount = (e) => {
    e.preventDefault();
    createUserWithEmailAndPassword(auth, newUser.email, newUser.password)
      .then((userCredential) => {
        const { user } = userCredential;
        updateProfile(auth.currentUser, {
          displayName: `${newUser.userName}`,
        })
          .then(() => console.log(user.displayName))
          .catch((err) => console.log(err));
        onSetLoggedInUser({
          ...loggedInUser,
          id: user.uid,
          userName: user.displayName,
        });
        axios
          .post("/user", { ...newUser })
          .then(
            setNewUser({
              email: "",
              password: "",
              userName: "",
            })
          )
          .catch((err) => console.log(err));
        navigate("/");
      })
      .catch((err) => console.log(err));
  };
  return (
    <div className="auth">
      <h2 className="auth-subheading">Create an Account</h2>
      <form className="auth-form">
        <label htmlFor="create-account_email-input">
          Email
          <input
            className="auth-form_input"
            onChange={handleChange}
            value={newUser.email}
            type="email"
            name="email"
            id="create-account_email-input"
          />
        </label>
        <label htmlFor="create-account_password-input">
          Password
          <input
            className="auth-form_input"
            onChange={handleChange}
            value={newUser.password}
            type="password"
            name="password"
            id="create-account_password-input"
          />
        </label>
        <label htmlFor="create-account_username-input">
          Username
          <input
            className="auth-form_input"
            onChange={handleChange}
            value={newUser.userName}
            type="userName"
            name="userName"
            id="create-account_username-input"
          />
        </label>
        <label htmlFor="create-account_firstname-input">
          First Name
          <input
            className="auth-form_input"
            onChange={handleChange}
            value={newUser.firstName}
            type="firstName"
            name="firstName"
            id="create-account_firstname-input"
          />
        </label>
        <label htmlFor="create-account_lastname-input">
          Last Name
          <input
            className="auth-form_input"
            onChange={handleChange}
            value={newUser.lastName}
            type="lastName"
            name="lastName"
            id="create-account_lastname-input"
          />
        </label>
        <label htmlFor="create-account_jobrole-input">
          Job Role
          <input
            className="auth-form_input"
            onChange={handleChange}
            value={newUser.jobRole}
            type="jobRole"
            name="jobRole"
            id="create-account_jobrole-input"
          />
        </label>
        <button
          id="create-user_button"
          type="button"
          onClick={handleCreateAccount}
        >
          Submit
        </button>
      </form>
    </div>
  );
};

export default CreateAccount;
