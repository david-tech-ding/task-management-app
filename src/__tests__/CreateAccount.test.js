import React from "react";
import axios from "axios";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { MemoryRouter as Router } from "react-router-dom";
import CreateAccount from "../components/CreateAccount";
import { createUserWithEmailAndPassword } from "firebase/auth";

jest.mock("../config/firebase", () => {
  return {
    auth: {
      currentUser: {
        updateProfile: jest.fn(),
      },
    },
  };
});

jest.mock("firebase/auth", () => {
  return {
    createUserWithEmailAndPassword: jest.fn(() =>
      Promise.resolve({ user: { displayName: "testUser" } })
    ),
  };
});

jest.mock("axios", () => ({
  post: jest.fn(),
  get: jest.fn(() => Promise.resolve({ data: [] })),
}));

describe("CreateAccount component", () => {
  it("renders correctly", () => {
    const { asFragment } = render(
      <Router>
        <CreateAccount />
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("updates state when input fields change", () => {
    render(
      <Router>
        <CreateAccount />
      </Router>
    );
    const emailInput = screen.getByLabelText("Email");
    const passwordInput = screen.getByLabelText("Password");
    const usernameInput = screen.getByLabelText("Username");

    fireEvent.change(emailInput, { target: { value: "test@example.com" } });
    fireEvent.change(passwordInput, { target: { value: "password123" } });
    fireEvent.change(usernameInput, { target: { value: "testUser123" } });

    expect(emailInput.value).toBe("test@example.com");
    expect(passwordInput.value).toBe("password123");
    expect(usernameInput.value).toBe("testUser123");
  });

  it("submits the form correctly", async () => {
    render(
      <Router>
        <CreateAccount />
      </Router>
    );
    const emailInput = screen.getByLabelText("Email");
    const passwordInput = screen.getByLabelText("Password");
    const usernameInput = screen.getByLabelText("Username");
    const submitButton = screen.getByText("Submit");

    fireEvent.change(emailInput, { target: { value: "test@example.com" } });
    fireEvent.change(passwordInput, { target: { value: "password123" } });
    fireEvent.change(usernameInput, { target: { value: "testUser123" } });

    axios.post.mockResolvedValue({});
    createUserWithEmailAndPassword.mockResolvedValue({ user: {} });

    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(createUserWithEmailAndPassword).toHaveBeenCalledWith(
        expect.any(Object),
        "test@example.com",
        "password123"
      );
    });
    await waitFor(() => {
      expect(axios.post).toHaveBeenCalledWith("/user", {
        email: "test@example.com",
        password: "password123",
        userName: "testUser123",
      });
    });
  });
});
