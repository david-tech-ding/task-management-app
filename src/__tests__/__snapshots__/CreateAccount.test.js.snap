// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateAccount component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="auth"
  >
    <h2
      class="auth-subheading"
    >
      Create an Account
    </h2>
    <form
      class="auth-form"
    >
      <label
        for="create-account_email-input"
      >
        Email
        <input
          class="auth-form_input"
          id="create-account_email-input"
          name="email"
          type="email"
          value=""
        />
      </label>
      <label
        for="create-account_password-input"
      >
        Password
        <input
          class="auth-form_input"
          id="create-account_password-input"
          name="password"
          type="password"
          value=""
        />
      </label>
      <label
        for="create-account_username-input"
      >
        Username
        <input
          class="auth-form_input"
          id="create-account_username-input"
          name="userName"
          type="userName"
          value=""
        />
      </label>
      <button
        id="create-user_button"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
